op-challenger/challenger/traversal.go

```diff
@@ -0,0 +1,247 @@
+		return err
+	})
+	return sub, err
+}
+
+// Quit quits the log traversal.
+func (l *logTraversal) Quit() {
+	l.quit <- struct{}{}
+	l.started = false
+}
+
+// Start starts the log traversal.
+func (l *logTraversal) Start(ctx context.Context, handleLog func(*types.Log) error) error {
+	if l.started {
+		return ErrAlreadyStarted
+	}
+	headers := make(chan *types.Header)
+	sub, err := l.subscribeNewHead(ctx, headers)
+	if err != nil {
+		l.log.Error("Failed to subscribe to new heads", "err", err)
+		return err
+	}
+	l.started = true
+	go l.onNewHead(ctx, headers, sub, handleLog)
+	return nil
+}
+
+// Started returns true if the log traversal has started.
+func (l *logTraversal) Started() bool {
+	return l.started
+}
+
+// onNewHead handles a new [types.Header].
+func (l *logTraversal) onNewHead(ctx context.Context, headers chan *types.Header, sub ethereum.Subscription, handleLog func(*types.Log) error) {
+	for {
+		select {
+		case <-l.quit:
+			l.log.Info("Stopping log traversal: received quit signal")
+			sub.Unsubscribe()
+			return
+		case header := <-headers:
+			l.log.Info("Received new head", "number", header.Number)
+			l.dispatchNewHead(ctx, header, handleLog, true)
+		}
+	}
+}
+
+// spawnCatchup spawns a new goroutine to "catchup" for missed/skipped blocks.
+func (l *logTraversal) spawnCatchup(ctx context.Context, start *big.Int, end *big.Int, handleLog func(*types.Log) error) {
+	for {
+		// Break if we've caught up (start > end).
+		if start.Cmp(end) == 1 {
+			l.log.Info("Traversal caught up", "start", start, "end", end)
+			return
+		}
+
+		info, err := l.fetchBlockInfoByNumber(ctx, start.Uint64())
+		if err != nil {
+			l.log.Error("Failed to fetch block", "err", err)
+			return
+		}
+
+		receipts, err := l.fetchBlockReceipts(ctx, info.Hash())
+		if err != nil {
+			l.log.Error("Failed to fetch receipts", "err", err)
+			return
+		}
+		for _, receipt := range receipts {
+			for _, log := range receipt.Logs {
+				err := handleLog(log)
+				if err != nil {
+					l.log.Error("Failed to handle log", "err", err)
+					return
+				}
+			}
+		}
+
+		// Increment to the next block to catch up to
+		start = start.Add(start, big.NewInt(1))
+	}
+}
+
+// updateBlockNumber updates the last block number with a mutex lock.
+func (l *logTraversal) updateBlockNumber(blockNumber *big.Int) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	l.lastBlockNumber = blockNumber
+}
+
+// dispatchNewHead dispatches a new head.
+func (l *logTraversal) dispatchNewHead(ctx context.Context, header *types.Header, handleLog func(*types.Log) error, allowCatchup bool) {
+	info, err := l.fetchBlockInfoByHash(ctx, header.Hash())
+	if err != nil {
+		l.log.Error("Failed to fetch block", "err", err)
+		return
+	}
+	expectedBlockNumber := l.lastBlockNumber.Add(l.lastBlockNumber, big.NewInt(1))
+	currentBlockNumber := big.NewInt(int64(info.NumberU64()))
+	if l.lastBlockNumber.Cmp(big.NewInt(0)) != 0 && currentBlockNumber.Cmp(expectedBlockNumber) == 1 {
```

Comment:
So we're just starting from whatever the first new block we receive after start up is? That will mean we fail to process any blocks that arrive while we're restarting/offline.

Reply:
Right, changed back to feeding a block number into the constructor